enhancements/ingress/gateway-api-with-cluster-ingress-operator.md

```diff
@@ -0,0 +1,434 @@
+    tls:
+      mode: Terminate
+      certificateRefs:
+      - name: gwapi-wildcard
+    allowedRoutes:
+      namespaces:
+        from: All
+```
+[1]: The requirement to specify `selector.istio` is an Istio-ism that needs to be investigated. \
+[2]: The requirement to specify `gatewayClassName: istio ` is another Istio-ism that is addressed by \
+     [istio/istio#38134](https://github.com/istio/istio/issues/38134) and \
+     [istio/istio#41884](https://github.com/istio/istio/pull/41884). \
+[3]: The requirement to specify the "istio-ingressgateway" service's DNS name \
+     in `spec.addresses` is another Istio-ism that needs to be discussed.
+
+Then a project admin can attach an HTTPRoute to the Gateway.
+
+_"As a cluster admin, I want to configure ingress using Gateway API, and I want
+to enable application owners to configure mesh using the Istio API, and I want
+these two things to work harmoniously."_
+
+The goals of this enhancement include that ingress and mesh use a unified
+control-plane (with a single Istiod instance).  The cluster admin should be able
+to enable mesh and subsequently enable ingress or vice versa without needing to
+reconfigure the mesh or ingress, respectively, in the subsequent step.
+
+_"As a project admin, I want to configure ingress to my application using an
+HTTPRoute, the same way I would use an OpenShift Route."_
+
+Once the cluster admin has configured a Gateway as described in the earlier user
+story, the project admin can define an HTTPRoute that attaches to the Gateway.
+For example:
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: HTTPRoute
+metadata:
+  name: http
+  namespace: example-app
+spec:
+  parentRefs:
+  - name: gateway
+    namespace: openshift-ingress
+  hostnames: ["test.gwapi.example.com"]
+  rules:
+  - backendRefs:
+    - name: example-app
+      port: 8080
+```
+
+This example exposes the "example-app" service, using port 8080 of the service.
+In this example, the HTTPRoute specifies a host name that matches both listeners
+on the Gateway, so the application is exposed using both HTTP and HTTPS.  When a
+client connects to the application using HTTPS, the default certificate that the
+cluster admin specified on the Gateway is used.  The Envoy proxy terminates the
+TLS connection and forwards the HTTP connection to the "example-app" service.
+
+### Goals
+
+* Automate the installation of Istio/Envoy by means of OpenShift Service Mesh.
+* Enable cluster admins to configure ingress using Gateway and GatewayClass.
+* Enable project admins to configure ingress to an application using HTTPRoute.
+* Allow cleartext HTTP and edge-terminated HTTPS traffic.
+* Manage a LoadBalancer-type service and wildcard DNS for ingress.
+* Use a unified control-plane (i.e. a single Istiod deployment) for ingress and mesh.
+* Work harmoniously with OpenShift Service Mesh for service mesh use-cases.
+* Provide a default security policy, which the cluster admin may modify.
+
+### Non-Goals
```

Comment:
> Should we add non-goals for things we are planning to do later?

Yes, if a reasonable person would wonder whether it might be part of the enhancement, then the goals and non-goals should specify whether or not it is.  

>     * Convert Route to Gateway API

I'll add this as a non-goal.  

>     * Ingress Canary for Gateway API

I'll add this too.  

>     * Any secret/cert management implications? (C-I-O) does a lot of cert management type things (like default cert and rotating).

Does the following already cover it?

> OpenShift does not automatically generate a default certificate for a Gateway; the cluster admin must do this.

Reply:
Sounds good.

> > Does the following already cover it?
> 
> OpenShift does not automatically generate a default certificate for a Gateway; the cluster admin must do this.

Yea I think that generally covers it.